src/relax/distributed/transform/lower_global_view_to_local_view.cc

```diff
@@ -0,0 +1,442 @@
+    if (new_block->name_hint == "root") {
+      new_block->alloc_buffers.insert(new_block->alloc_buffers.end(),
+                                      allocated_buffer_under_root.begin(),
+                                      allocated_buffer_under_root.end());
+    }
+    new_block->body = DistBufferReplacer::BufferReplace(block->body, buffer_map);
+    return Block(new_block);
+  }
+
+  void AddAllReduceBlock(std::string reduce_kind) { add_allreduce_kind_ = reduce_kind; }
+
+  Stmt VisitStmt_(const BlockRealizeNode* op) final {
+    BlockRealize realize = Downcast<BlockRealize>(StmtExprMutator::VisitStmt_(op));
+
+    for (int i = 0; i < static_cast<int>(realize->iter_values.size()); i++) {
+      PrimExpr iter_value = realize->iter_values[i];
+      IterVar iter_var = realize->block->iter_vars[i];
+      if (!iter_var_shards_.count(iter_var->var)) {
+        continue;
+      }
+      ICHECK(iter_value.as<VarNode>());
+      loop_var_shards_[Downcast<Var>(iter_value)] = iter_var_shards_[iter_var->var];
+    }
+    return realize;
+  }
+
+  Stmt VisitStmt_(const ForNode* op) final {
+    For new_loop = Downcast<For>(StmtExprMutator::VisitStmt_(op));
+    if (loop_var_shards_.count(op->loop_var)) {
+      int shard = loop_var_shards_[op->loop_var];
+      if (shard > 1) {
+        arith::Analyzer analyzer;
+        ICHECK(analyzer.CanProve(floormod(new_loop->extent, shard) == 0));
+        return For(new_loop->loop_var, new_loop->min, floordiv(new_loop->extent, shard),
+                   new_loop->kind, new_loop->body, new_loop->thread_binding, new_loop->annotations);
+      }
+    }
+    return new_loop;
+  }
+
+  std::unordered_map<Var, int, ObjectPtrHash, ObjectPtrEqual> iter_var_shards_;
+  std::unordered_map<Var, int, ObjectPtrHash, ObjectPtrEqual> loop_var_shards_;
+  Array<Buffer> allocated_buffer_under_root;
+  BufferAxisGraphExtractor extractor_;
+  std::vector<ShardingSpec> sharding_specs_;
+  std::unordered_map<Buffer, DimShard, ObjectPtrHash, ObjectPtrEqual> buffer_shards_;
+  std::string add_allreduce_kind_;
+};
+
+}  // namespace tir
+}  // namespace tvm
+
+namespace tvm {
+namespace relax {
+namespace distributed {
+
+class LowerTIRToLocalView : public ExprMutator {
+ public:
+  explicit LowerTIRToLocalView(IRModule mod) : ExprMutator(mod) {}
+
+  IRModule Lower() {
+    auto mod = builder_->GetContextIRModule();
+    for (const auto& [gv, base_func] : mod->functions) {
+      const auto* func_ = base_func.as<FunctionNode>();
+      if (func_ == nullptr || !IsDistIRFunc(GetRef<Function>(func_))) {
+        continue;
+      }
+      Expr new_func_body = this->VisitExpr(func_->body);
+      ObjectPtr<FunctionNode> new_func = make_object<FunctionNode>(*func_);
+      new_func->body = new_func_body;
+      builder_->UpdateFunction(gv, Function(new_func));
+    }
+    return builder_->GetContextIRModule();
+  }
+
+ private:
+  inline Array<DTensorStructInfo> ExtractDTensorStructInfo(Var var) {
```

Comment:
could it be a nested tuple (eg: `tup(dtensor, tup(dtensor, dtensor))`)? not sure if this case is disallowed

Reply:
good catch. My current assumption is to disallow nested tuple. And we can surely support it if we find it's necessary to do so.